cirq/circuits/circuit.py

```diff
@@ -68,74 +73,159 @@ class Circuit:
+        '''
+        from cirq.circuits import FrozenCircuit
+        if isinstance(self, FrozenCircuit):
+            return self
+        return FrozenCircuit(self,
+                             strategy=InsertStrategy.EARLIEST,
+                             device=self.device)
+
+    def unfreeze(self) -> 'cirq.Circuit':
+        '''Creates a Circuit from this circuit.
+
+        If 'self' is a Circuit, the original object is returned.
+        '''
+        if isinstance(self, Circuit):
+            return self
+        return Circuit(self,
+                       strategy=InsertStrategy.EARLIEST,
+                       device=self.device)
 
     def __bool__(self):
-        return bool(self._moments)
+        return bool(self.moments)
 
     def __eq__(self, other):
         if not isinstance(other, type(self)):
             return NotImplemented
-        return self._moments == other._moments and self._device == other._device
+        return self.moments == other.moments and self.device == other.device
 
     def _approx_eq_(self, other: Any, atol: Union[int, float]) -> bool:
         """See `cirq.protocols.SupportsApproximateEquality`."""
         if not isinstance(other, type(self)):
             return NotImplemented
         return cirq.protocols.approx_eq(
-            self._moments,
-            other._moments,
-            atol=atol
-        ) and self._device == other._device
+            self.moments, other.moments,
+            atol=atol) and self.device == other.device
 
     def __ne__(self, other) -> bool:
         return not self == other
 
     def __len__(self) -> int:
-        return len(self._moments)
+        return len(self.moments)
 
     def __iter__(self) -> Iterator['cirq.Moment']:
-        return iter(self._moments)
+        return iter(self.moments)
 
     def _decompose_(self) -> 'cirq.OP_TREE':
         """See `cirq.SupportsDecompose`."""
         return self.all_operations()
 
+    # pylint: disable=function-redefined
+    @overload
+    def __getitem__(self, key: int) -> 'cirq.Moment':
+        pass
+
+    @overload
+    def __getitem__(self, key: Tuple[int, 'cirq.Qid']) -> 'cirq.Operation':
+        pass
+
+    @overload
+    def __getitem__(self,
+                    key: Tuple[int, Iterable['cirq.Qid']]) -> 'cirq.Moment':
+        pass
+
+    @overload
+    def __getitem__(self, key: slice):
+        pass
+
+    @overload
+    def __getitem__(self, key: Tuple[slice, 'cirq.Qid']):
+        pass
+
+    @overload
+    def __getitem__(self, key: Tuple[slice, Iterable['cirq.Qid']]):
+        pass
+
+    def __getitem__(self, key):
+        if hasattr(key, '__index__'):
+            return self.moments[key]
+        if isinstance(key, tuple):
+            if len(key) != 2:
+                raise ValueError('If key is tuple, it must be a pair.')
+            moment_idx, qubit_idx = key
+            # qubit_idx - Qid or Iterable[Qid].
+            return self.moments[moment_idx][qubit_idx]
```

Comment:
Does this actually work if `moment_idx` is a slice? It seems we need to slice the qubits on each moment in the sequence `self.moments[moment_idx]` and then reassemble into the correct circuit type.

Edit: ok, I see that `Circuit` and `FrozenCircuit` implement some of the `__getitem__` logic and then fallback to this `__getitem__` implementation. I think it might be more clear to put all the logic here and call into helper methods that must be implemented on the implementing classes as needed, e.g. we build all the moments here and then call `self._with_sliced_moments(moments)` to construct the new instance.